Split locals and args

nbhuiyan opened the pull request:
Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Split Args/Locals handling with new instructions

Function arguments and locals are no longer handled as
just locals.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

rwy7 commented on b9/src/ExecutionContext.cpp:
```diff
@@ -223,7 +228,15 @@ void ExecutionContext::doPushFromVar(StackElement *args, Immediate offset) {
+  args[offset] = stack_.pop();
+}
+
+void ExecutionContext::doPushFromArg(StackElement *args, Immediate offset){
```
formatting

rwy7 commented on b9/src/ExecutionContext.cpp:
```diff
@@ -223,7 +228,15 @@ void ExecutionContext::doPushFromVar(StackElement *args, Immediate offset) {
+  stack_.push(args[offset]);
+}
+
+void ExecutionContext::doPopIntoArg(StackElement *args, Immediate offset){
```
formatting

rwy7 commented on b9/src/ExecutionContext.cpp:
```diff
@@ -119,8 +119,7 @@ StackElement ExecutionContext::interpret(const std::size_t functionIndex) {
       case OpCode::POP_INTO_VAR:
-        // TODO bad name, push or pop?
-        doPushIntoVar(args, instructionPointer->immediate());
+        doPopIntoVar(args, instructionPointer->immediate());
```
*sigh* what a silly mistake 🐱
thanks for fixing.

rwy7 commented on b9/include/b9/instructions.hpp:
```diff
@@ -70,6 +70,12 @@ enum class OpCode : RawOpCode {
+
+  //Pop into a parameter
+  POP_INTO_ARG = 0x19,
+
```
I know this is a lot of tedious work, but can we have the push/pop from arg come right after (or before) the var instructions?

rwy7 commented on b9/src/MethodBuilder.cpp:
```diff
@@ -335,11 +335,13 @@ bool MethodBuilder::generateILForBytecode(
       if (nextBytecodeBuilder)
         builder->AddFallThroughBuilder(nextBytecodeBuilder);
       break;
+    case OpCode::POP_INTO_ARG:
```
OK, looking at the frontend, the parameters and locals each use a different symbol table. This means that args and params are indexed differently; arg[0] is not the same as local[0]. But here, it looks like arg[0] and local[0] are aliased. Is this a bug? There's the same issue in the interpreter.

rwy7 commented on b9run/main.cpp:
```diff
@@ -35,7 +35,7 @@ static const char* usage =
   b9::Config b9;
   const char* moduleName = "";
-  const char* mainFunction = "<body>";
+  const char* mainFunction = "b9main";
```
why did we change this?

rwy7 reviewed (changes requested):
IIUC, there are some issues with arguments and parameters being overlayed on the same stack slots.

On top of that, I would really like to get our terminology consistent, and I think this PR is a good place to start:

1) parameters vs arguments: I suggest going with param or parameter everywhere. It's closer to the JS standard and esprima terminology. parameter used to be an slightly overloaded term, but instruction parameters were renamed to immediates.

2) vars vs locals vs registers: suggesting we go with local. Originally I was pushing "register", but that word is probably confusing.

nbhuiyan committed:
Fix parameters and locals indexing

A fatal oversight with how stores/loads happen with
args and vars have been fixed by using the necessary
offsetting.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Improve opcode order for better grouping

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Make params and locals terminology more consistent

Previously, params were referred to as both arguments and
parameters. Locals were referred to as locals, vars and registers.
The selected terminology is closer to the JavaScript standard and
Esprima terminology. A consistent terminology makes things easier
for everyone.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Improve handling of locals and params for the JIT

load/storeVal handles both locals and params, with the necessary
offsetting required for locals done by load/storeLocalIndex.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
@rwy0717 Ready for another review. There is a merge conflict that I will fix after the changes have been approved.

rwy7 commented on b9/include/b9/instructions.hpp:
```diff
@@ -255,8 +263,8 @@ inline std::ostream &operator<<(std::ostream &out, Instruction i) {
-    case OpCode::PUSH_FROM_VAR:
-    case OpCode::POP_INTO_VAR:
+    case OpCode::PUSH_FROM_LOCAL:
+    case OpCode::POP_INTO_LOCAL:
```
Missing push/pop param

rwy7 replied:
nevermind... how did I miss it :p

nbhuiyan replied:
I missed it too at first and only recently pushed that :P

nbhuiyan committed:
Explicitly list push_from/pop_into param as 1 imm instructions

They would treated as 1 imm instructions anyway due to the
default case.

Also includes minor fixes in MethodBuilder.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
@rwy0717 The merge conflicts have been fixed, and the CI build has passed. I think it is ready to be merged.

rwy7 commented:
LGTM. Now that we differentiate between locals and arguments, we should take another look at how the JIT feature flags (passparam, lazy-vm-state) affect how locals and parameters behave. Thanks @nbhuiyan. closes #154.
